### ISSUE TITLE
[Test-Proxy] Some headers not auto-splitting, causing recording mismatches

### ISSUE THREAD

**Issue by scbedd** [Azure/azure-sdk-tools#3058]
@Mohit-Chakraborty is seeing matching failures when converting `containerregistry` to use the test-proxy. While he has unblocked himself by ignoring this specific header during matching, it still highlights a problem.

[Right here](https://github.com/Azure/azure-sdk-tools/blob/f46fb45cf9603d3d705ca832ef8c0e562bf7c307/tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs#L425) we create an entry that will matched against when running in `playback` mode. 

The key issue here is that `ASP.NET` is treating the header value as `string[]{ "application/vnd.oci.image.manifest.v1+json, application/json" }` instead of what I would actually expect, which would be to automatically split by `,` and result in `string[]{ "application/vnd.oci.image.manifest.v1+json", "application/json" }`.

Given that specific input string, the headers just don't split consistently. Because of _that_, the test proxy is simply storing that value as the 0th index in the header values array. I believe we probably record on `.NET 6.0` which is the form with the space in it. If we had recorded in .NET 4.6.1, this problem would simply be reversed.

This issue covers the investigation to understand why we aren't getting the expected values array. 

If it is a bug in ASP.NET in general, we will need to re-factor how we treat headers across the board to one where we explicitly look at the joined raw value.

**Comment by scbedd** [comment]
Oh. I don't think this is a bug. `,` definitely doesn't auto-split, in which case this is just expected behavior because we're passing two different values. Going to find the LoC in asp.net code and drop a reference. This is probably going to end up a `wont fix` unless I understand something different.

**Comment by scbedd** [comment]
I am going to close this issue as  `won't fix`, I don't believe this is an issue with the test-proxy. We don't split up that header value as discussed above, and we're simply sending _different values_ across the framework values. I don't believe that we should be applying logic to detect 

> oh that space is immaterial because it is in the `Accept` header, you can ignore if it's there or not there

@JoshLove-msft @Mohit-Chakraborty feel free to challenge my understanding of the problem above!

**Comment by Mohit-Chakraborty** [comment]
Thanks @scbedd for the investigation.
Could you explain what you mean by this -
> we're simply sending different values across the different .NET frameworks.

Should we try to standardize on the format? What would the recommendation be for writing tests targeting different .NET frameworks? Perhaps have a way to keep separate test session records?

**Comment by JoshLove-msft** [comment]
We intentionally normalize the Accept header to handle the case of different formats. It seems this is just one more format that we need to handle when normalizing the headers.

**Comment by scbedd** [comment]
@JoshLove-msft this won't be normalized right? Headers are auto split based on `;`, NOT `,`. This is a single header value.

**Comment by scbedd** [comment]
Hey @Mohit-Chakraborty , laying out the exact details of the differences.

So when this test runs on `.NET 5.0/6.0`, you send along this value for the `Accept` header: `application/vnd.oci.image.manifest.v1+json, application/json`

On the failing `.net 4.6.1`, this same test sends this value for `Accept` header: `"application/vnd.oci.image.manifest.v1+json,application/json"`

Notice that there isn't a space there? 

For test authors, it's really what you have in your recording. It looks like you MADE the recording on `.NET 6.0`, which is why the space is present.

**Comment by JoshLove-msft** [comment]
> @JoshLove-msft this won't be normalized right? Headers are auto split based on `;`, NOT `,`. This is a single header value.
> 
> EDIT: oh you're talking about .NET side. nevermind.

I know they are not being normalized now, but I don't understand why we wouldn't normalize them. I thought comma separated values are supported - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept

**Comment by JoshLove-msft** [comment]
My guess is that the code here - https://github.com/Azure/azure-sdk-tools/blob/main/tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordMatcher.cs#L255 doesn't work anymore now that we are after the transport layer. The "normalization" that occurs in this method still needs to happen but are starting point is no longer an array with separate content types, but rather a comma-delimited list of content-types.

**Comment by scbedd** [comment]
@JoshLove-msft even if it did this header is a single value, and that normalization wouldn't do anything anyway. `Accept` special cases on `,`. Most headers are on `;`.

**Comment by JoshLove-msft** [comment]
The normalization is specifically for Accept and Content-Type - https://github.com/Azure/azure-sdk-tools/blob/f46fb45cf9603d3d705ca832ef8c0e562bf7c307/tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordMatcher.cs#L24-L25

**Comment by Mohit-Chakraborty** [comment]
For context, [the SDK code](https://github.com/Azure/azure-sdk-for-net/blob/7cd48c766d4e71300345dee8ac8a2a2958b02a5c/sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Generated/ContainerRegistryRestClient.cs#L97-L101) is -
```
            if (accept != null)
            {
                request.Headers.Add("accept", accept);
            }
            request.Headers.Add("Accept", "application/json");
```

**Comment by JoshLove-msft** [comment]
> @JoshLove-msft even if it did this header is a single value, and that normalization wouldn't do anything anyway. `Accept` special cases on `,`. Most headers are on `;`.

Yes, I agree the normalization method is currently not doing anything because it expects the header values to be in separate elements of an array. The method then joins these elements into a comma separated string. The problem is that we are starting off with a comma separated string as the only element of the array, but we still need to do the normalization part that adds the correct whitespace.

**Comment by scbedd** [comment]
Right, because it's not two header values. It's a single header value.

**Comment by JoshLove-msft** [comment]
> Accept",

I doubt it will make a difference but any reason there is different casing used?

**Comment by Mohit-Chakraborty** [comment]
> > Accept",
> 
> I doubt it will make a difference but any reason there is different casing used?

That gets unified. But, otherwise it is odd, and a bug in CodeGen.
